File: delay_action_cpp/src/action_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "action_interfaces/srv/string.hpp"
#include "std_msgs/msg/string.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>

using namespace std::chrono_literals;

class ActionClient : public rclcpp::Node
{
  public:
    ActionClient() : Node("action_client")
    {
      client_ = this->create_client<action_interfaces::srv::String>("action");
      action_subscriber_ = this->create_subscription<std_msgs::msg::String>("new_action", 10, std::bind(&ActionClient::action_cb, this, std::placeholders::_1));

      RCLCPP_INFO(this->get_logger(), "Action Client Ready. Listening on 'new_action' topic");
    }

    void action_cb(const std_msgs::msg::String &msg) {
      RCLCPP_INFO(this->get_logger(), "New action received from subscriber. Action: '%s'", msg.data.c_str());

      this->send_request(msg.data);
    }

    void send_request(std::string input)
    {
      auto request = std::make_shared<action_interfaces::srv::String::Request>();
      request->data = input;

      //[FIX]:  Change position of this check
      while(!client_->wait_for_service(1s)) {
        if (!rclcpp::ok()) {
          RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
          return;
        }
        RCLCPP_INFO(this->get_logger(), "action service not available, waiting again...");
      }

      auto result = client_->async_send_request(
        request,
        std::bind(&ActionClient::handle_response, this, std::placeholders::_1));
    }

    void handle_response(rclcpp::Client<action_interfaces::srv::String>::SharedFuture future)
    {
      auto response = future.get();
      RCLCPP_INFO(this->get_logger(), "Output: %s", response->output.c_str());
    }
  
  private:
    rclcpp::Client<action_interfaces::srv::String>::SharedPtr client_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr action_subscriber_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  rclcpp::spin(std::make_shared<ActionClient>());

  rclcpp::shutdown();
  return 0;
}

File: delay_action_py/delay_action_py/action_client.py
from action_interfaces.srv import String as AString
from std_msgs.msg import String

import rclpy
from rclpy.node import Node

class ActionClient(Node):
    def __init__(self):
        super().__init__("action_client")
        self.client_ = self.create_client(AString, "action")
        
        while not self.client_.wait_for_service(1.0):
            if not rclpy.ok():
                self.get_logger().info("Interrupted while waiting for the service. Exiting.")
                return
            self.get_logger().info("action service not available, waiting again...")
        
        self.subscriber_ = self.create_subscription(String, "new_action", self.action_sub_cb, 10)

        self.get_logger().info("Action Client Ready. Listening on 'new_action' topic")

    
    def action_sub_cb(self, msg: String):
        self.get_logger().info("New action received from subscriber. Action: '%s'" % msg.data)
        
        self.send_request(msg.data)

    def send_request(self, input: str):
        req = AString.Request()
        req.data = input

        future = self.client_.call_async(req)
        future.add_done_callback(self.handle_response)

    def handle_response(self, future):
        response = future.result()
        self.get_logger().info("Output: %s" % response.output)
    
def main(args=None):
    rclpy.init(args=args)

    action_client = ActionClient()

    rclpy.spin(action_client)
    
    action_client.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()


        
